### PR TITLE
Crud decorator refactor (fixes #12)

### LINKED ISSUE #12: Refactor crud decorator

### PR DESCRIPTION
**Pull request by leplatrem**
(Alex's edit of empty description)

This pull request does:

- [x] Refactor the `@crud` decorator into a `register_resource` function and a `ViewSet` class;
- [x] Add unit tests for these two;
- [x] Document the new additions;
- [x] Create a cliquet subclass of `cornice.Service` to avoid cluttering the cornice namespace;
- [x] Handle `readonly` and `readwrite` permissions;
- [x] Handle CORS in a better way than before;
- [x] Make the functional tests pass.

### CODE REVIEW

**Comment by ametaireau** [comment]
Okay, now we're starting to get somewhere!

I've updated the description to show where we're at. I would like feedback on these areas specifically:

- There is room for improvement about how tests are done. I only have one class for each thing I want to test, I probably could create some other `TestClass`, but which?;
- I've not put the CORS definition in the new `register_resource` function, because non-resources views would have missed it. Is there any other way to do this?;
- The design of the solution;
- Wording.

**Comment by ametaireau** [comment]
Ran 682 tests in 11.105s, FAILED (errors=2)

I currently have the following two tests failing:

```
======================================================================
ERROR: test_modify_with_invalid_body_returns_400 (cliquet.tests.resource.test_views.InvalidBodyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/alexis/dev/github.com/cliquet/cliquet/tests/resource/test_views.py", line 216, in test_modify_with_invalid_body_returns_400
    status=400)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 405, in patch
    content_type=content_type)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 735, in _gen_request
    expect_errors=expect_errors)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 631, in do_request
    self._check_status(status, res)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 666, in _check_status
    "Bad response: %s (not %s)", res_status, status)
AppError: Bad response: 500 Internal Server Error (not 400)
-------------------- >> begin captured logging << --------------------
venusian: INFO: "POST  /v0/mushrooms" 201 (1 ms) request.summary lang=None; auth_type=FxA; resource_name=mushroom; uid=fxa_bob; errno=None; resource_timestamp=1432836798613; agent=None; time=2015-05-28T20:13:18
venusian: ERROR: "PATCH /v0/mushrooms/1e0e4aa9-8d02-405c-aa3b-b535e4be9c8d" ? (? ms) Expected object or value lang=None; auth_type=FxA; exception=Traceback (most recent call last):
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/tweens.py", line 21, in excview_tween
    response = handler(request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/router.py", line 163, in handle_request
    response = view_callable(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 596, in __call__
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 329, in attr_view
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 305, in predicate_wrapper
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 245, in _secured_view
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 355, in rendered_view
    result = view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 501, in _requestonly_view
    response = view(request)
  File "/home/alexis/dev/github.com/cornice/cornice/service.py", line 568, in wrapper
    response = view_()
  File "/home/alexis/dev/github.com/cliquet/cliquet/resource.py", line 472, in patch
    changes = self.request.json
  File "build/bdist.linux-x86_64/egg/webob/request.py", line 725, in _json_body__get
    return json.loads(self.body.decode(self.charset))
ValueError: Expected object or value; resource_name=mushroom; uid=fxa_bob; errno=None; agent=None; resource_timestamp=1432836798614
venusian: INFO: "PATCH /v0/mushrooms/1e0e4aa9-8d02-405c-aa3b-b535e4be9c8d" 500 (3 ms) request.summary lang=None; auth_type=FxA; resource_name=mushroom; uid=fxa_bob; errno=999; resource_timestamp=1432836798614; agent=None; time=2015-05-28T20:13:18
--------------------- >> end captured logging << ---------------------

======================================================================
```

And 

```
ERROR: test_modify_with_invalid_uft8_returns_400 (cliquet.tests.resource.test_views.InvalidBodyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/alexis/dev/github.com/cliquet/cliquet/tests/resource/test_views.py", line 237, in test_modify_with_invalid_uft8_returns_400
    status=400)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 405, in patch
    content_type=content_type)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 735, in _gen_request
    expect_errors=expect_errors)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 631, in do_request
    self._check_status(status, res)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/webtest/app.py", line 666, in _check_status
    "Bad response: %s (not %s)", res_status, status)
AppError: Bad response: 500 Internal Server Error (not 400)
-------------------- >> begin captured logging << --------------------
venusian: INFO: "POST  /v0/mushrooms" 201 (2 ms) request.summary lang=None; auth_type=FxA; resource_name=mushroom; uid=fxa_bob; errno=None; resource_timestamp=1432836798621; agent=None; time=2015-05-28T20:13:18
venusian: ERROR: "PATCH /v0/mushrooms/c2ba621f-b5ab-43c3-990d-902c73d62655" ? (? ms) Unexpected character in unicode escape sequence when decoding 'string' lang=None; auth_type=FxA; exception=Traceback (most recent call last):
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/tweens.py", line 21, in excview_tween
    response = handler(request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/router.py", line 163, in handle_request
    response = view_callable(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 596, in __call__
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 329, in attr_view
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 305, in predicate_wrapper
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 245, in _secured_view
    return view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 355, in rendered_view
    result = view(context, request)
  File "/home/alexis/dev/github.com/cliquet/.venv/local/lib/python2.7/site-packages/pyramid-1.5.4-py2.7.egg/pyramid/config/views.py", line 501, in _requestonly_view
    response = view(request)
  File "/home/alexis/dev/github.com/cornice/cornice/service.py", line 568, in wrapper
    response = view_()
  File "/home/alexis/dev/github.com/cliquet/cliquet/resource.py", line 472, in patch
    changes = self.request.json
  File "build/bdist.linux-x86_64/egg/webob/request.py", line 725, in _json_body__get
    return json.loads(self.body.decode(self.charset))
ValueError: Unexpected character in unicode escape sequence when decoding 'string'; resource_name=mushroom; uid=fxa_bob; errno=None; agent=None; resource_timestamp=1432836798622
venusian: INFO: "PATCH /v0/mushrooms/c2ba621f-b5ab-43c3-990d-902c73d62655" 500 (1 ms) request.summary lang=None; auth_type=FxA; resource_name=mushroom; uid=fxa_bob; errno=999; resource_timestamp=1432836798622; agent=None; time=2015-05-28T20:13:18
--------------------- >> end captured logging << ---------------------
```

File: cliquet/resource.py [hunk 1]
```diff
@@ -19,37 +22,210 @@
...
+        self.update(**kwargs)
+        self.record_arguments = functools.partial(
+                                        self.get_arguments, 'record')
+        self.collection_arguments = functools.partial(
+                                        self.get_arguments, 'collection')
+
+    def update(self, **kwargs):
+        self.__dict__.update(**kwargs)
+
+    def get_arguments(self, typ_, resource, method):
```
Comment (leplatrem) [inline cliquet/resource.py]: ``typ_`` is not very explicit and could refer to a class...  ``endpoint`` ?

File: cliquet/resource.py [hunk 2]
```diff
@@ -19,37 +22,210 @@
...
-    :meth:`~cornice:cornice.resource.resource` decorator.
+    def get_name(self, resource):
+        """Returns the name of the resource.
+        """
+        if 'name' in self.__dict__:
+            name = self.__dict__['name']
+        elif hasattr(resource, 'name') and not callable(resource.name):
+            name = resource.name
+        else:
+            name = resource.__name__.lower()
```
Comment (leplatrem) [inline cliquet/resource.py]: This was refactored massively in #282

File: cliquet/resource.py [hunk 3]
```diff
@@ -19,37 +22,210 @@
...
+            return '%s_%s' % (typ_, method.lower())
 
-    This decorator accepts the same parameters as the :rtd:`Cornice <cornice>`
-    :meth:`~cornice:cornice.resource.resource` decorator.
+    def get_name(self, resource):
+        """Returns the name of the resource.
+        """
+        if 'name' in self.__dict__:
+            name = self.__dict__['name']
+        elif hasattr(resource, 'name') and not callable(resource.name):
```
Comment (leplatrem) [inline cliquet/resource.py]: When I tried with Kinto code, this test about callable didn't work

File: cliquet/resource.py [hunk 4]
```diff
@@ -19,37 +22,210 @@
...
-    .. code-block:: python
+        return name
 
-            from cliquet import resource
+    def get_service_name(self, typ_, resource):
+        """Returns the name of the service, depending a given type and
+        resource.
+        """
+        name = self.get_name(resource)
+        return '-'.join((name, typ_))
```
Comment (leplatrem) [inline cliquet/resource.py]: This could be a ViewSet class attribute with string placeholders (like description)

File: cliquet/resource.py (hunk 3 above)
Comment (ametaireau) [inline cliquet/resource.py]: That's strange, I added it to make it work for kinto, actually. Can you be more specific about what didn't work?

File: cliquet/resource.py (hunk 1 above)
Comment (ametaireau) [inline cliquet/resource.py]: I agree, but `endpoint` doesn't seem that explicit either. `endpoint_type`?

File: cliquet/resource.py (hunk 2 above)
Comment (ametaireau) [inline cliquet/resource.py]: So, what's the new way of handling this?

Comment (leplatrem) [inline cliquet/resource.py]: ``name`` is now an attribute of ``Collection`` and can be changed at any time with ``resource.collection.name = 'XX'``
```

File: cliquet/__init__.py [hunk 5]
```diff
@@ -79,6 +79,14 @@
 }
 
 
+class Service(CorniceService):
+    """Subclass of the default cornice service.
+
+    This is useful in order to attach specific behaviours without monkey
+    patching the default cornice service (which would impact other uses of it)
+    """
```
Comment (leplatrem) [inline cliquet/__init__.py]: But is it since we don't extend anything ?

File: cliquet/resource.py [hunk 6]
```diff
@@ -19,37 +22,214 @@
...
+
+    def get_view(self, endpoint_type, method):
+        """Returns the view location for the given type and method.
+
+        For collections, this will be "collection_{method|lower}
+        For records, this will be "{method|lower}.
+        """
+        if endpoint_type == 'record':
+            return method.lower()
+        else:
```
Comment (leplatrem) [inline cliquet/resource.py]: ``else`` is superflous ?

File: cliquet/tests/test_resource.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
+from mock import sentinel, MagicMock, patch
```
Comment (leplatrem) [inline cliquet/tests/test_resource.py]: Maybe a matter of taste, but I would keep them as ``mock.sentinel``, ``mock.MagicMock``, ``mock.patch``

File: cliquet/tests/test_resource.py [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+        self.assertEquals(viewset.collection_path,
+                          sentinel.collection_path)
+
+    def test_default_arguments_are_copied_before_being_returned(self):
+        original_arguments = {}
+        viewset = ViewSet(
+            collection_get_arguments=original_arguments,
+            validate_schema_for=()
+        )
+        arguments = viewset.collection_arguments(sentinel.resource, 'GET')
```
Comment (leplatrem) [inline cliquet/tests/test_resource.py]: you could ``assertNotEquals(arguments, original_arguments)`` here

File: cliquet/resource.py [hunk 9]
```diff
@@ -19,37 +22,214 @@
...
+        self.update(**kwargs)
+        self.record_arguments = functools.partial(
+                                        self.get_arguments, 'record')
+        self.collection_arguments = functools.partial(
+                                        self.get_arguments, 'collection')
+
+    def update(self, **kwargs):
+        self.__dict__.update(**kwargs)
+
+    def get_arguments(self, endpoint_type, resource, method):
```
Comment (leplatrem) [inline cliquet/resource.py]: ``get_view_args()`` ?)

File: cliquet/tests/test_resource.py [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+    def test_schema_is_added_when_method_matches(self, patched):
+        viewset = ViewSet(
+            validate_schema_for=('GET', )
+        )
+        resource = MagicMock()
+        arguments = viewset.collection_arguments(resource, 'GET')
+        self.assertEquals(arguments['schema'], resource.mapping)
+
+    @patch('cliquet.resource.CorniceSchema')
+    def test_schema_is_added_when_method_is_uppercase(self, patched):
```
Comment (leplatrem) [inline cliquet/tests/test_resource.py]: when uppercase method matches

File: cliquet/tests/test_resource.py [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+    def test_get_service_name_doesnt_use_callable_as_a_name(self):
+        viewset = ViewSet()
+        resource = MagicMock()
+        resource.name = lambda x: 'should not be called'
+        resource.__name__ = "FakeName"
+        self.assertEquals(
+            viewset.get_service_name('record', resource),
+            'fakename-record')
+
+    def test_get_view_permission_returns_readwrite_by_default(self):
```
Comment (leplatrem) [inline cliquet/tests/test_resource.py]: You could group permission tests together into a ``ViewPermissionTest``

File: cliquet/tests/test_resource.py [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+            arguments,
+            {
+                'cors_headers': sentinel.default_cors_headers,
+                'error_handler': sentinel.default_record_error_handler,
+                'cors_origins': sentinel.record_get_cors_origin,
+                'permission': 'readonly'
+            }
+        )
+
+    def test_service_arguments_arent_inherited_by_record_arguments(self):
```
Comment (leplatrem) [inline cliquet/tests/test_resource.py]: You could group arguments tests together into a ``CorniceArgumentsTest``, and put the one they have in common inside ``setUp``

File: cliquet/__init__.py (hunk 5 above)
Comment (ametaireau) [inline cliquet/__init__.py]: We are extending `CorniceService` and reusing it in all non-resource based views.

**Comment by ametaireau** [comment]
Updated.

**Pull request merged**